src/middleware.rs

```diff
@@ -0,0 +1,169 @@
+/// fn my_middleware(req: Request, next: MiddlewareNext) -> Result<Response, Error> {
+///     // do middleware things
+///
+///     // continue the middleware chain
+///     next.handle(req)
+/// }
+/// ```
+///
+/// # Adding headers
+///
+/// A common use case is to add headers to the outgoing request. Here an example of how.
+///
+/// ```
+/// # #[cfg(feature = "json")]
+/// # fn main() -> Result<(), ureq::Error> {
+/// # use ureq::{Request, Response, MiddlewareNext, Error};
+/// # ureq::is_test(true);
+/// fn my_middleware(req: Request, next: MiddlewareNext) -> Result<Response, Error> {
+///     // set my bespoke header and continue the chain
+///     next.handle(req.set("X-My-Header", "value_42"))
+/// }
+///
+/// let agent = ureq::builder()
+///     .middleware(my_middleware)
+///     .build();
+///
+/// let result: serde_json::Value =
+///     agent.get("http://httpbin.org/headers").call()?.into_json()?;
+///
+/// assert_eq!(&result["headers"]["X-My-Header"], "value_42");
+///
+/// # Ok(()) }
+/// # #[cfg(not(feature = "json"))]
+/// # fn main() {}
+/// ```
+///
+/// # State
+///
+/// To maintain state between middleware invocations, we need to do something more elaborate than
+/// the simple `fn` and implement the `Middleware` trait directly.
+///
+/// ## Example with mutex lock
+///
+/// In the `examples` directory there is an additional example `count-bytes.rs` which uses
+/// a mutex lock like shown below.
+///
+/// ```
+/// # use ureq::{Request, Response, Middleware, MiddlewareNext, Error};
+/// # use std::sync::{Arc, Mutex};
+/// struct MyState {
+///     // whatever is needed
+/// }
+///
+/// struct MyMiddleware(Arc<Mutex<MyState>>);
+///
+/// impl Middleware for MyMiddleware {
+///     fn handle(&self, request: Request, next: MiddlewareNext) -> Result<Response, Error> {
+///         // These extra brackets ensures we release the Mutex lock before continuing the
+///         // chain. There could also be scenarios where we want to maintain the lock through
+///         // the invocation, which would block other requests from proceeding concurrently
+///         // through the middleware.
+///         {
+///             let mut state = self.0.lock().unwrap();
+///             // do stuff with state
+///         }
+///
+///         // continue middleware chain
+///         next.handle(request)
+///     }
+/// }
+/// ```
+///
+/// ## Example with atomic
+///
+/// This example shows how we can increase a counter for each request going
+/// through the agent.
+///
+/// ```
+/// # fn main() -> Result<(), ureq::Error> {
+/// # ureq::is_test(true);
+/// use ureq::{Request, Response, Middleware, MiddlewareNext, Error};
+/// use std::sync::atomic::{AtomicU64, Ordering};
+/// use std::sync::Arc;
+///
+/// // Middleware that stores a counter state. This example uses an AtomicU64
+/// // since the middleware is potentially shared by multiple threads running
+/// // requests at the same time.
+/// struct MyCounter(Arc<AtomicU64>);
```

Comment:
Probably this makes a bit more sense as:

```
struct MyCounter(AtomicU64);
```

And then down below we construct `Arc::new(MyCounter::new())`, and give one close to the Agent and keep one clone for ourselves. In other words, `Arc` isn't part of the internal guarantees of MyCounter, it's just there so we can share ownership with Agent (which requires a 'static lifetime bound).

Reply:
Alright. I was trying to mirror how the example is doing it where an arc/mutex protects an inner state.

Not sure what is less confusing to a beginner? 🤔

Reply:
Now that you mention it, I also think `MyMiddleware` should be `struct MyMiddleware(Mutex<MyState>)` and we should pass `Arc<MyMiddleware>` to the Agent.

Reply:
I think this is not clear separation of concerns.

A common pattern for shared mutable state is `Arc<Mutex<T>>`. A middleware working with such a state is then naturally `MyMiddleware(Arc<Mutex<T>>)`.

That we want an additional `Arc<dyn Middleware>` to facilitate cloning of the middleware is a different issue. Sure. Technically these two could just be one single `Arc`, but I think it confuses two ideas.